api/src/main/java/org/eclipse/microprofile/metrics/MetricID.java

```diff
@@ -99,14 +89,31 @@
      */
     private final Map<String, String> tags = new TreeMap<String, String>();
 
+    /**
+     * Constructs a MetricID with the given metric name, global tags defined
+     * by the specified {@link GlobalTagsProvider}, and additional {@link Tag}s.
+     *
+     * @param name the name of the metric
+     * @param globalTagsProvider the provider instance to retrieve the global tags from;
+     *                           can be {@code null}
+     * @param tags the (non-global) tags associated with this metric
+     */
+    public MetricID(String name, GlobalTagsProvider globalTagsProvider, Tag... tags) {
+        this.name = name;
+        if (globalTagsProvider != null) {
+            globalTagsProvider.getGlobalTags().forEach(this::addTag);
+        }
+        addTags(tags);
+    }
+
```

Comment:
This method is essentially saying you can ignore the globally configured tags and instead use what the caller says are the global tags for this metric.  But then for all metrics that don't come from this caller you'll have a different set of global tags.  I'm not sure I understand the use case for this.

Reply:
It simply guards against the case when there may not be a global tags provider, but if you think we should require one, I'm ok with that too.

I guess that probably makes sense, considering that any spec-compliant implementation will have to support at least config-defined global tags. I'll make the change and document that `GlobalTagsProvider` cannot be `null`.

Reply:
Done.

Reply:
Given the global tags are meant for the implementation to provide, why don't we make the `DefaultGlobalTagsProvider` look up the impl of `GlobalTagsProvider` using service loader and use it to fill in the metrics.  That would avoid us having to expose the `GlobalTagsProvider` in the `MetricsId` API, which I think is better as app developers shouldn't be concerned with that concept.

Reply:
App developers may not be concerned, but MP Metrics implementors certainly should be, as it allows them to optimize things.

Otherwise, everyone has to rely on the default implementation, which is no better than the current solution. The original stated goal of #506 was to allow implementations to have better control over the implementation of the spec, but somehow the discussion has devolved from that into which constructors we need and should have on `MetricID` :-(

In any case, if you feel strongly about not exposing new constructor to everyone, we can make it protected to allow implementation to use it, but that would force each implementation to extend `MetricId` for no other reason, which I believe is not really necessary.

Reply:
The implementation will provide the `GlobalTagProviders` available through ServiceLoader, so that's how they will be picked up. I don't think we need a special constructor that takes a `GlobalTagProvider` argument. Certainly applications don't need it (I hope we all agree on that), but I don't think the implementation needs that either.

Reply:
@aseovic, I agree with your point that the goal is to give more flexibility to implementations to implement the spec their way.  My concern is just that we're exposing things that only an implementor should be concerned about in the API that app developers will use.  That worries me because it would confuse app developers that don't know what to do with this parameter.  I think @jmartisk and I are on a similar point -- that GlobalTagProvider is something only the impl should ever see, which is why I like the idea of using Service Loader to load GlobalTagProvider within the DefaultGlobalTagsProvider.  That would avoid GlobalTagProvider being something on the classes that app writers interact with.

Reply:
As far as I understand the main pain point of current global tags implementation are the bad performance implications and the lack of control for implementers to solve this more elegantly. In this context I fully agree that such logic never should have been in the API. Adding the service loader indirection does allow more control but also opens a new problem vector. If service locator is used implementers have to assume it is used by users which will only give them the option to hope that the user does implement the caching of reading the config which is almost unlikely. It also opens up possibility for users to make the implementation non conforming. Giving the user both a configuration key and an abstraction feels unnecessary duplication and burden both for users to understand and for implementers to handle the complexity this introduces. I made a suggestion https://github.com/eclipse/microprofile-metrics/pull/560#issuecomment-607154941 to entirely move global tags out of `MetricID` class which I feel removes all the pain points and gives all control back to implementers while users are not bothered with the detail of global tags except that they can use the configuration to set them.

Reply:
> If service locator is used implementers have to assume it is used by users

I don't think we need to be concerned with that. We can just put in the spec that this is meant for implementations, and bundling a provider directly in your application results in undefined, non portable behavior

Reply:
> We can just put in the spec that this is meant for implementations

If this solution goes forward I'd consider this a very important point to include.